{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SPARQLWrapper import SPARQLWrapper, JSON\n",
    "import pandas as pd\n",
    "import urllib\n",
    "import requests\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ongoing problems:\n",
    "when I start trying to query sparql for a bunch of articles that contain 'open science' in the abstract, I keep getting this 502 proxy error. At the moment, my best guess is that this is some kind of timeout. When I go to the ma-graph endpoint and paste in the same text, but specify the timeout thing to be long, I get data back (though perhaps not everything I should). My next stage is, I think, to start looking into timeouts.\n",
    "\n",
    "If that fails, I can download the dumps. I presently have 1tb of storage on this machine. prom has about 2.5tb and could maybe get more on prom.\n",
    "\n",
    "# Data description\n",
    "Here, I'm pulling data out of a sparql database that includes information from microsoft academic knowledge graph. The website for this data is here: http://ma-graph.org/\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sparql = SPARQLWrapper('http://ma-graph.org/sparql')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "sparql.setQuery(\"\"\"\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX magp: <http://ma-graph.org/property/>\n",
    "PREFIX mag: <http://ma-graph.org/>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "PREFIX foaf: <http://xmlns.com/foaf/0.1/>\n",
    "PREFIX fabio: <http://purl.org/spar/fabio/>\n",
    "PREFIX org: <http://www.w3.org/ns/org#>\n",
    "PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "PREFIX cito: <http://purl.org/spar/cito/>\n",
    "PREFIX c4o: <http://purl.org/spar/c4o/>\n",
    "PREFIX prism: <http://prismstandard.org/namespaces/1.2/basic/>\n",
    "PREFIX datacite: <http://purl.org/spar/datacite/>\n",
    "\n",
    "SELECT ?nihauthname ?otherauths ?papertitle ?discname ?dt ?doi ?abst\n",
    "WHERE {\n",
    "?inst rdf:type <http://ma-graph.org/class/Affiliation> .\n",
    "?inst foaf:name \"National Institutes of Health\"^^xsd:string .\n",
    "?auth org:memberOf ?inst .\n",
    "?paper dcterms:creator ?auth .\n",
    "?auth foaf:name ?nihauthname .\n",
    "?paper dcterms:creator ?allauths .\n",
    "?allauths foaf:name ?otherauths .\n",
    "?paper dcterms:title ?papertitle .\n",
    "?paper prism:publicationDate ?dt .\n",
    "?paper fabio:hasDiscipline ?disc .\n",
    "?paper datacite:doi ?doi .\n",
    "?paper dcterms:abstract ?abst .\n",
    "?disc foaf:name ?discname .\n",
    "}\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "#?paper cito:hasCitedEntity ?cit .\n",
    "#?cit c4o:hasContext ?citcont ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "sparql.setReturnFormat(JSON)\n",
    "results = sparql.query().convert()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1048576"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(results['results']['bindings'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'nihauthname': {'type': 'typed-literal',\n",
       "  'datatype': 'http://www.w3.org/2001/XMLSchema#string',\n",
       "  'value': 'Scott R. Budihas'},\n",
       " 'otherauths': {'type': 'typed-literal',\n",
       "  'datatype': 'http://www.w3.org/2001/XMLSchema#string',\n",
       "  'value': 'Michael A. Parniak'},\n",
       " 'papertitle': {'type': 'typed-literal',\n",
       "  'datatype': 'http://www.w3.org/2001/XMLSchema#string',\n",
       "  'value': 'A capillary electrophoretic assay for ribonuclease H activity.'},\n",
       " 'discname': {'type': 'typed-literal',\n",
       "  'datatype': 'http://www.w3.org/2001/XMLSchema#string',\n",
       "  'value': 'Biochemistry'},\n",
       " 'dt': {'type': 'typed-literal',\n",
       "  'datatype': 'http://www.w3.org/2001/XMLSchema#date',\n",
       "  'value': '2004-08-01'},\n",
       " 'doi': {'type': 'typed-literal',\n",
       "  'datatype': 'http://www.w3.org/2001/XMLSchema#string',\n",
       "  'value': '10.1016/j.ab.2004.05.017'},\n",
       " 'abst': {'type': 'typed-literal',\n",
       "  'datatype': 'http://www.w3.org/2001/XMLSchema#string',\n",
       "  'value': 'Abstract A capillary electrophoretic assay was developed to measure the ribonuclease (RNase) H activity of human immunodeficiency virus (HIV) type 1 reverse transcriptase. Cleavage of a fluorescein-labeled RNA\\x13DNA heteroduplex was monitored by capillary electrophoresis. This new assay was used as a secondary assay to confirm hits from a high-throughput screening program. Since autofluorescent compounds in samples migrated differently from both substrate and product in most cases, the assay was extremely robust for assaying enzymatic inhibition of such samples, in contrast to a simple well-based approach. The assay was broadly applicable to other RNases H, specifically those from human, Escherichia coli , and HIV-2, although product profiles varied for each enzyme.'}}"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results['results']['bindings'][10000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "outdat = []\n",
    "for i in results['results']['bindings']:\n",
    "    outdat.append([i['papertitle']['value'], i['authname']['value'], \n",
    "                   i['discname']['value'], i['dt']['value'], i['doi']['value']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "dat = pd.DataFrame(outdat, columns = ['title', 'author', 'field', 'date', 'doi'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>author</th>\n",
       "      <th>field</th>\n",
       "      <th>date</th>\n",
       "      <th>doi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Concetta Castilletti</td>\n",
       "      <td>Ebola virus</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Concetta Castilletti</td>\n",
       "      <td>Virology</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Concetta Castilletti</td>\n",
       "      <td>Subclade</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Concetta Castilletti</td>\n",
       "      <td>Outbreak</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Concetta Castilletti</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Concetta Castilletti</td>\n",
       "      <td>Biology</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160224</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Silvia Meschi</td>\n",
       "      <td>Ebola virus</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160225</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Silvia Meschi</td>\n",
       "      <td>Virology</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160226</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Silvia Meschi</td>\n",
       "      <td>Subclade</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160227</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Silvia Meschi</td>\n",
       "      <td>Outbreak</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160228</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Silvia Meschi</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160229</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Silvia Meschi</td>\n",
       "      <td>Biology</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391311</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Antonino Di Caro</td>\n",
       "      <td>Ebola virus</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391312</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Antonino Di Caro</td>\n",
       "      <td>Virology</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391313</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Antonino Di Caro</td>\n",
       "      <td>Subclade</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391314</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Antonino Di Caro</td>\n",
       "      <td>Outbreak</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391315</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Antonino Di Caro</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391316</th>\n",
       "      <td>Molecular Signature of the Ebola Virus Associa...</td>\n",
       "      <td>Antonino Di Caro</td>\n",
       "      <td>Biology</td>\n",
       "      <td>2015-10-29</td>\n",
       "      <td>10.1128/genomeA.01093-15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    title  \\\n",
       "0       Molecular Signature of the Ebola Virus Associa...   \n",
       "1       Molecular Signature of the Ebola Virus Associa...   \n",
       "2       Molecular Signature of the Ebola Virus Associa...   \n",
       "3       Molecular Signature of the Ebola Virus Associa...   \n",
       "4       Molecular Signature of the Ebola Virus Associa...   \n",
       "5       Molecular Signature of the Ebola Virus Associa...   \n",
       "160224  Molecular Signature of the Ebola Virus Associa...   \n",
       "160225  Molecular Signature of the Ebola Virus Associa...   \n",
       "160226  Molecular Signature of the Ebola Virus Associa...   \n",
       "160227  Molecular Signature of the Ebola Virus Associa...   \n",
       "160228  Molecular Signature of the Ebola Virus Associa...   \n",
       "160229  Molecular Signature of the Ebola Virus Associa...   \n",
       "391311  Molecular Signature of the Ebola Virus Associa...   \n",
       "391312  Molecular Signature of the Ebola Virus Associa...   \n",
       "391313  Molecular Signature of the Ebola Virus Associa...   \n",
       "391314  Molecular Signature of the Ebola Virus Associa...   \n",
       "391315  Molecular Signature of the Ebola Virus Associa...   \n",
       "391316  Molecular Signature of the Ebola Virus Associa...   \n",
       "\n",
       "                      author           field        date  \\\n",
       "0       Concetta Castilletti     Ebola virus  2015-10-29   \n",
       "1       Concetta Castilletti        Virology  2015-10-29   \n",
       "2       Concetta Castilletti        Subclade  2015-10-29   \n",
       "3       Concetta Castilletti        Outbreak  2015-10-29   \n",
       "4       Concetta Castilletti  Bioinformatics  2015-10-29   \n",
       "5       Concetta Castilletti         Biology  2015-10-29   \n",
       "160224         Silvia Meschi     Ebola virus  2015-10-29   \n",
       "160225         Silvia Meschi        Virology  2015-10-29   \n",
       "160226         Silvia Meschi        Subclade  2015-10-29   \n",
       "160227         Silvia Meschi        Outbreak  2015-10-29   \n",
       "160228         Silvia Meschi  Bioinformatics  2015-10-29   \n",
       "160229         Silvia Meschi         Biology  2015-10-29   \n",
       "391311      Antonino Di Caro     Ebola virus  2015-10-29   \n",
       "391312      Antonino Di Caro        Virology  2015-10-29   \n",
       "391313      Antonino Di Caro        Subclade  2015-10-29   \n",
       "391314      Antonino Di Caro        Outbreak  2015-10-29   \n",
       "391315      Antonino Di Caro  Bioinformatics  2015-10-29   \n",
       "391316      Antonino Di Caro         Biology  2015-10-29   \n",
       "\n",
       "                             doi  \n",
       "0       10.1128/genomeA.01093-15  \n",
       "1       10.1128/genomeA.01093-15  \n",
       "2       10.1128/genomeA.01093-15  \n",
       "3       10.1128/genomeA.01093-15  \n",
       "4       10.1128/genomeA.01093-15  \n",
       "5       10.1128/genomeA.01093-15  \n",
       "160224  10.1128/genomeA.01093-15  \n",
       "160225  10.1128/genomeA.01093-15  \n",
       "160226  10.1128/genomeA.01093-15  \n",
       "160227  10.1128/genomeA.01093-15  \n",
       "160228  10.1128/genomeA.01093-15  \n",
       "160229  10.1128/genomeA.01093-15  \n",
       "391311  10.1128/genomeA.01093-15  \n",
       "391312  10.1128/genomeA.01093-15  \n",
       "391313  10.1128/genomeA.01093-15  \n",
       "391314  10.1128/genomeA.01093-15  \n",
       "391315  10.1128/genomeA.01093-15  \n",
       "391316  10.1128/genomeA.01093-15  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat[dat.doi=='10.1128/genomeA.01093-15']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>field</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Biology</th>\n",
       "      <td>56725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Medicine</th>\n",
       "      <td>41832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Molecular biology</th>\n",
       "      <td>23265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Immunology</th>\n",
       "      <td>22135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Internal medicine</th>\n",
       "      <td>19022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Virology</th>\n",
       "      <td>15039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Biochemistry</th>\n",
       "      <td>14901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Endocrinology</th>\n",
       "      <td>13882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genetics</th>\n",
       "      <td>13696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cancer</th>\n",
       "      <td>9490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chemistry</th>\n",
       "      <td>8870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathology</th>\n",
       "      <td>8655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cancer research</th>\n",
       "      <td>7490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Diabetes mellitus</th>\n",
       "      <td>6095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <td>5912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Psychology</th>\n",
       "      <td>5773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Virus</th>\n",
       "      <td>5252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Physical therapy</th>\n",
       "      <td>4983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Immune system</th>\n",
       "      <td>4951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Antibody</th>\n",
       "      <td>4856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Disease</th>\n",
       "      <td>4786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cardiology</th>\n",
       "      <td>4611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell biology</th>\n",
       "      <td>4440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Antigen</th>\n",
       "      <td>4089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Surgery</th>\n",
       "      <td>4022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Receptor</th>\n",
       "      <td>3610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Anatomy</th>\n",
       "      <td>3413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Anesthesia</th>\n",
       "      <td>2886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T cell</th>\n",
       "      <td>2751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation</th>\n",
       "      <td>2588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Linear molecular geometry</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Linear no-threshold model</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Linear phase</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Linear predictor function</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ligneous conjunctivitis</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lightness</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Light-harvesting complex</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Light-emitting diode</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Levinthal&amp;#039;s paradox</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Levoamphetamine</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Levodopa-induced dyskinesia</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Levorphanol</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lewis acid catalysis</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Leydig cell hypoplasia</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Liberalism</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Libman\u0013Sacks endocarditis</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Library management</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lichtheimia corymbifera</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lidoflazine</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lien</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Life skills</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Life support</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lifestyle management</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lift (force)</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ligature</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Light beam</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Light chain deposition disease</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Light field</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Light scattering by particles</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>xDNA</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29873 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                title\n",
       "field                                \n",
       "Biology                         56725\n",
       "Medicine                        41832\n",
       "Molecular biology               23265\n",
       "Immunology                      22135\n",
       "Internal medicine               19022\n",
       "Virology                        15039\n",
       "Biochemistry                    14901\n",
       "Endocrinology                   13882\n",
       "Genetics                        13696\n",
       "Cancer                           9490\n",
       "Chemistry                        8870\n",
       "Pathology                        8655\n",
       "Cancer research                  7490\n",
       "Diabetes mellitus                6095\n",
       "Gene                             5912\n",
       "Psychology                       5773\n",
       "Virus                            5252\n",
       "Physical therapy                 4983\n",
       "Immune system                    4951\n",
       "Antibody                         4856\n",
       "Disease                          4786\n",
       "Cardiology                       4611\n",
       "Cell biology                     4440\n",
       "Antigen                          4089\n",
       "Surgery                          4022\n",
       "Receptor                         3610\n",
       "Anatomy                          3413\n",
       "Anesthesia                       2886\n",
       "T cell                           2751\n",
       "Mutation                         2588\n",
       "...                               ...\n",
       "Linear molecular geometry           1\n",
       "Linear no-threshold model           1\n",
       "Linear phase                        1\n",
       "Linear predictor function           1\n",
       "Ligneous conjunctivitis             1\n",
       "Lightness                           1\n",
       "Light-harvesting complex            1\n",
       "Light-emitting diode                1\n",
       "Levinthal&#039;s paradox            1\n",
       "Levoamphetamine                     1\n",
       "Levodopa-induced dyskinesia         1\n",
       "Levorphanol                         1\n",
       "Lewis acid catalysis                1\n",
       "Leydig cell hypoplasia              1\n",
       "Liberalism                          1\n",
       "Libman\u0013Sacks endocarditis           1\n",
       "Library management                  1\n",
       "Lichtheimia corymbifera             1\n",
       "Lidoflazine                         1\n",
       "Lien                                1\n",
       "Life skills                         1\n",
       "Life support                        1\n",
       "Lifestyle management                1\n",
       "Lift (force)                        1\n",
       "Ligature                            1\n",
       "Light beam                          1\n",
       "Light chain deposition disease      1\n",
       "Light field                         1\n",
       "Light scattering by particles       1\n",
       "xDNA                                1\n",
       "\n",
       "[29873 rows x 1 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat.groupby('field').count()[['title']].sort_values('title', ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# URL presence\n",
    "\n",
    "To what extent have open-data repositories been adopted? Have some repositories been adopted more quickly than others? Does this vary by field?\n",
    "\n",
    "These questions have bearing on our understanding of the effects of open-data, and open science more generally. In principle, open science is designed to facilitate the participation of a wide range of people in the scientific community. If groups that have a disproportionate share of resources consistently publish their likely resource-intensive datasets, then these valuable datasets can be mined by a wider range of individuals for scientific insights. \n",
    "\n",
    "To this end, I will examine how frequently URL's to some prominent open science repositories are present in papers as urls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "sparql.setQuery(\"\"\"\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX magp: <http://ma-graph.org/property/>\n",
    "PREFIX mag: <http://ma-graph.org/>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "PREFIX foaf: <http://xmlns.com/foaf/0.1/>\n",
    "PREFIX fabio: <http://purl.org/spar/fabio/>\n",
    "PREFIX org: <http://www.w3.org/ns/org#>\n",
    "PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "PREFIX cito: <http://purl.org/spar/cito/>\n",
    "PREFIX datacite: <http://purl.org/spar/datacite/>\n",
    "PREFIX prism: <http://prismstandard.org/namespaces/1.2/basic/>\n",
    "PREFIX c4o: <http://purl.org/spar/c4o/>\n",
    "\n",
    "SELECT ?paper ?paptitle ?abst ?dt ?ln ?auth ?doi ?estCites\n",
    "WHERE {\n",
    "?paper rdf:type <http://ma-graph.org/class/Paper> .\n",
    "?paper dcterms:title ?paptitle .\n",
    "?paper dcterms:abstract ?abst .\n",
    "?paper prism:publicationDate ?dt .\n",
    "?paper dcterms:creator ?auth .\n",
    "?paper datacite:doi ?doi .\n",
    "?paper dcterms:language ?ln .\n",
    "?paper dcterms:creator ?auth .\n",
    "}\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "sparql.setReturnFormat(JSON)\n",
    "results = sparql.query().convert() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1048576"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(results['results']['bindings'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'paper': {'type': 'uri', 'value': 'http://ma-graph.org/entity/4524729'},\n",
       " 'paptitle': {'type': 'typed-literal',\n",
       "  'datatype': 'http://www.w3.org/2001/XMLSchema#string',\n",
       "  'value': 'Fiber Optics Sensors'},\n",
       " 'abst': {'type': 'typed-literal',\n",
       "  'datatype': 'http://www.w3.org/2001/XMLSchema#string',\n",
       "  'value': 'Fiber optic technology started in 1970\\x19s, for long distance telecommunications, and it has experienced an exponential growth during the last four decades. Sensing applications are a small spin-off from this technology, taking profit of developments done for optoelectronic components and concepts. Optical fibers are cylindrical dielectric waveguides for the propagation of the light, made out from high purity, low loss optical materials, usually silica (Optical fibers made from polymers and other transparent materials are also commercially available). The refractive index of the core (about 1.46 for silica) is slightly higher than surrounding material or cladding, due to the presence of dopants. Light is confined at the core, because optical rays arriving to the interface core/cladding with an angle higher than the total reflectance angle, as defined by the Snell law, will follow total reflections and will remain confined at the core. Only when the fiber is locally bent with a sharp radius, the light may escape.'},\n",
       " 'dt': {'type': 'typed-literal',\n",
       "  'datatype': 'http://www.w3.org/2001/XMLSchema#date',\n",
       "  'value': '2013-01-01'},\n",
       " 'ln': {'type': 'typed-literal',\n",
       "  'datatype': 'http://www.w3.org/2001/XMLSchema#language',\n",
       "  'value': 'en'},\n",
       " 'doi': {'type': 'typed-literal',\n",
       "  'datatype': 'http://www.w3.org/2001/XMLSchema#string',\n",
       "  'value': '10.1007/978-3-7091-1390-5_5'}}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results['results']['bindings'][2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "outdat = []\n",
    "for i in results['results']['bindings']:\n",
    "    outdat.append([i['paper']['value'], i['paptitle']['value'], \n",
    "                   i['abst']['value'], i['dt']['value'], i['ln']['value'],\n",
    "                   i['doi']['value']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "dat = pd.DataFrame(outdat, columns = ['paper', 'title', 'abstract', 'date',\n",
    "                                      'language', 'doi'])\n",
    "dat = dat[dat.lang=='en']\n",
    "dat['open_science'] = dat.abstract.str.contains('open science', case=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "dat.to_pickle('/home/riddleta/ac_knowl/output/testdat.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "sparql.setQuery(\"\"\"\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX magp: <http://ma-graph.org/property/>\n",
    "PREFIX mag: <http://ma-graph.org/>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "PREFIX foaf: <http://xmlns.com/foaf/0.1/>\n",
    "PREFIX fabio: <http://purl.org/spar/fabio/>\n",
    "PREFIX org: <http://www.w3.org/ns/org#>\n",
    "PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "PREFIX cito: <http://purl.org/spar/cito/>\n",
    "PREFIX datacite: <http://purl.org/spar/datacite/>\n",
    "PREFIX prism: <http://prismstandard.org/namespaces/1.2/basic/>\n",
    "PREFIX c4o: <http://purl.org/spar/c4o/>\n",
    "\n",
    "SELECT ?author ?dt ?cites ?papes ?authname ?rnk ?inst\n",
    "WHERE {\n",
    "?author rdf:type <http://ma-graph.org/class/Author> .\n",
    "}\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "sparql.setReturnFormat(JSON)\n",
    "results = sparql.query().convert() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1048576"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(results['results']['bindings'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "imagine a matrix of field x institution/author, with the cells of the matrix representing either counts of papers, or binary occurrences. What fields have a diversity of authors/institutions? what are more monopolistics? what are the consequences of the differences between these especially over time?\n",
    "\n",
    "is there an association between data sharing and market concentration (e.g. hh index)?\n",
    "\n",
    "also, look at what aaron clausett has been working on\n",
    "\n",
    "http://www.semantic-web-journal.net/system/files/swj1495.pdf\n",
    "\n",
    "group differences in \"tenure rates, competency evaluations, remuneration, allocation of research facilities, and\n",
    "grant competitions\" (way, larremore & clauset, 2016)\n",
    "\n",
    "\n",
    "plots to make:\n",
    "\n",
    "prevalency of links to various data-sharing platforms over time (see the [google spreadsheet](https://docs.google.com/spreadsheets/d/14GSgOppnvjS-pxGz1__fEBuj3KUNgKB7ni5nukHT8Ps/edit#gid=2016420688) for links)\n",
    "\n",
    "market concentration by...author? institution?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Aaron Clauset:\n",
    "https://arxiv.org/abs/1612.08228\n",
    "\n",
    "https://arxiv.org/abs/1804.02760\n",
    "\n",
    "https://arxiv.org/abs/1805.09966"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dani Bassett:\n",
    "https://arxiv.org/abs/1810.10534\n",
    "\n",
    "https://arxiv.org/abs/1804.06434\n",
    "\n",
    "https://www.sciencedirect.com/science/article/pii/S105381191830778X?via%3Dihub"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Jevin West:\n",
    "https://jevinwest.org/papers/West2013PLoSOne.pdf\n",
    "\n",
    "https://jevinwest.org/papers/West2014EconInquiry.pdf\n",
    "\n",
    "https://jevinwest.org/papers/Vilhena2014SocScience.pdf\n",
    "\n",
    "https://jevinwest.org/papers/Larsen2014HICSS.pdf\n",
    "\n",
    "https://jevinwest.org/papers/Portenoy2016WWW.pdf\n",
    "\n",
    "https://jevinwest.org/papers/Kim2017asa.pdf\n",
    "\n",
    "https://jevinwest.org/papers/Chow2017fisheries.pdf\n",
    "\n",
    "https://jevinwest.org/papers/King2017socius.pdf\n",
    "\n",
    "https://jevinwest.org/papers/Piwowar2018peerj.pdf\n",
    "\n",
    "\n",
    "# Michael Farber\n",
    "http://dbis.informatik.uni-freiburg.de/content/team/faerber/papers/arXiv_Dataset_BIR2019.pdf\n",
    "\n",
    "\n",
    "\n",
    "# Other:\n",
    "https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005962\n",
    "\n",
    "https://cdn.elifesciences.org/articles/32822/elife-32822-v2.pdf\n",
    "\n",
    "https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btz070/5305021\n",
    "\n",
    "https://link.springer.com/article/10.1007/s11192-012-0900-9\n",
    "\n",
    "https://link.springer.com/chapter/10.1007/978-3-319-67056-0_18\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5761855/\n",
    "\n",
    "https://onlinelibrary.wiley.com/doi/full/10.1002/asi.23266"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# datasets\n",
    "# arxiv (downloaded)\n",
    "# https://www.ncbi.nlm.nih.gov/research/bionlp/APIs/BioC-PMC/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pulling arxiv papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/Downloads/arxiv_v2.1/0704.0001.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-78fa68d43234>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/Downloads/arxiv_v2.1/0704.0001.txt'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'w'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/Downloads/arxiv_v2.1/0704.0001.txt'"
     ]
    }
   ],
   "source": [
    "f = open('/Downloads/arxiv_v2.1/0704.0001.txt','w') \n",
    "print(f) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get the pubmed papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/riddleta/ac_knowl/ac_knowl/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: read_table is deprecated, use read_csv instead.\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(598386, 4)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#bioc-pmc files\n",
    "files_to_get = pd.read_table('/home/riddleta/ac_knowl/data/filelist.txt', sep='\\s+')\n",
    "files_to_get.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1479862, 5)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oa_comm_use_files = pd.read_csv('/home/riddleta/ac_knowl/data/oa_comm_use_file_list.txt', \n",
    "                                sep='\\\\t', engine='python', header=None, skiprows=1)\n",
    "oa_comm_use_files.columns = ['File', 'ref', 'PMCID', 'PMID', 'license']\n",
    "oa_comm_use_files.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/riddleta/ac_knowl/ac_knowl/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: read_table is deprecated, use read_csv instead.\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(2408932, 5)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oa_files = pd.read_table('/home/riddleta/ac_knowl/data/oa_file_list.txt', \n",
    "                         sep='\\\\t', engine='python', header=None, skiprows=1)\n",
    "oa_files.columns = ['File', 'ref', 'PMCID', 'PMID', 'license']\n",
    "oa_files.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_ids = pd.concat([oa_files['PMCID'], oa_comm_use_files['PMCID'], files_to_get['PMCID']]).unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_useful_data(r):\n",
    "    git_hits = 0\n",
    "    osf_hits = 0\n",
    "    nda_hits = 0\n",
    "    open_neuro = 0\n",
    "    fmri = 0\n",
    "    res = 0\n",
    "    r_out = r.json()\n",
    "    date = r_out['date']\n",
    "    src = r_out['source']\n",
    "    try:\n",
    "        doi = r_out['documents'][0]['passages'][0]['infons']['article-id_doi']\n",
    "    except:\n",
    "        doi = None\n",
    "    try:\n",
    "        yr = r_out['documents'][0]['passages'][0]['infons']['year']\n",
    "    except:\n",
    "        yr = None\n",
    "    if yr not in ['2008', '2009', '2010', '2011', '2012', '2013', \n",
    "                  '2014', '2015', '2016', '2017', '2018', '2019']:\n",
    "        return(date, src, doi, git_hits, osf_hits, nda_hits, open_neuro, fmri, res, yr)\n",
    "    else:\n",
    "        for passage in r_out['documents'][0]['passages']:\n",
    "            if re.search(re.compile('github'), passage['text'].lower()):\n",
    "                git_hits = git_hits + 1\n",
    "            if re.search(re.compile('osf\\.io'), passage['text'].lower()):\n",
    "                osf_hits = osf_hits + 1\n",
    "            if re.search(re.compile('nda\\.nih\\.gov'), passage['text'].lower()):\n",
    "                nda_hits = nda_hits + 1\n",
    "            if re.search(re.compile('openneuro'), passage['text'].lower()):\n",
    "                open_neuro = open_neuro + 1\n",
    "            if re.search(re.compile('fmri'), passage['text'].lower()):\n",
    "                fmri = fmri + 1\n",
    "            if re.search(re.compile('results'), passage['text'].lower()):\n",
    "                res = res + 1\n",
    "        \n",
    "        return(date, src, doi, git_hits, osf_hits, nda_hits, open_neuro, fmri, res, yr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "git_links = []\n",
    "for i, j in enumerate(unique_ids):\n",
    "    pth = 'https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_json/' + j + '/unicode'\n",
    "    r = requests.get(pth)\n",
    "    git_links.append(return_useful_data(r) + (j, i))\n",
    "    if i==10000:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(git_links, columns=['date', 'src', 'doi', 'git_hits', \n",
    "                                      'osf_hits', 'nda_hits', 'open_neuro', \n",
    "                                      'fmri', 'res', 'yr', 'pmcid', 'idx'])\n",
    "df.to_csv('/home/riddleta/ac_knowl/output/bionlp_10k.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "git_links = []\n",
    "for i, j in enumerate(unique_ids[10000:]):\n",
    "    pth = 'https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_json/' + j + '/unicode'\n",
    "    r = requests.get(pth)\n",
    "    git_links.append(return_useful_data(r) + (j, i))\n",
    "    if i==15000:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(git_links, columns=['date', 'src', 'doi', 'git_hits', \n",
    "                                      'osf_hits', 'nda_hits', 'open_neuro', \n",
    "                                      'fmri', 'res', 'yr', 'pmcid', 'idx'])\n",
    "df.to_csv('/home/riddleta/ac_knowl/output/bionlp_02.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "git_links = []\n",
    "missed = []\n",
    "\n",
    "for i, j in enumerate(unique_ids[25000:]):\n",
    "    pth = 'https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_json/' + j + '/unicode'\n",
    "    try:\n",
    "        r = requests.get(pth)\n",
    "        git_links.append(return_useful_data(r) + (j, i))\n",
    "    except:\n",
    "        missed.append([i,j])\n",
    "    if i==20000:\n",
    "        break\n",
    "df = pd.DataFrame(git_links, columns=['date', 'src', 'doi', 'git_hits', \n",
    "                                      'osf_hits', 'nda_hits', 'open_neuro', \n",
    "                                      'fmri', 'res', 'yr', 'pmcid', 'idx'])\n",
    "df.to_csv('/home/riddleta/ac_knowl/output/bionlp_03.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "git_links = []\n",
    "missed = []\n",
    "\n",
    "for i, j in enumerate(unique_ids[45000:]):\n",
    "    pth = 'https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_json/' + j + '/unicode'\n",
    "    try:\n",
    "        r = requests.get(pth)\n",
    "        git_links.append(return_useful_data(r) + (j, i))\n",
    "    except:\n",
    "        missed.append([i,j])\n",
    "    if i==25000:\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>src</th>\n",
       "      <th>doi</th>\n",
       "      <th>git_hits</th>\n",
       "      <th>osf_hits</th>\n",
       "      <th>nda_hits</th>\n",
       "      <th>open_neuro</th>\n",
       "      <th>fmri</th>\n",
       "      <th>res</th>\n",
       "      <th>yr</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20181121</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1186/1743-8462-2-29</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005</td>\n",
       "      <td>PMC1345690</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20181121</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1186/1471-2334-5-116</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005</td>\n",
       "      <td>PMC1345691</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20181121</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1186/1471-2105-6-282</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005</td>\n",
       "      <td>PMC1345692</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20181121</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj431</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006</td>\n",
       "      <td>PMC1345693</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20181121</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj441</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006</td>\n",
       "      <td>PMC1345694</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>20181121</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj453</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006</td>\n",
       "      <td>PMC1345695</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>20181121</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj461</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006</td>\n",
       "      <td>PMC1345696</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>20181124</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj463</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006</td>\n",
       "      <td>PMC1345697</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>20181121</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj468</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006</td>\n",
       "      <td>PMC1345698</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>20181121</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gnj008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006</td>\n",
       "      <td>PMC1345699</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>20181112</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gnj010</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006</td>\n",
       "      <td>PMC1345700</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>20181112</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gnj012</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006</td>\n",
       "      <td>PMC1345701</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>20181112</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gnj015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006</td>\n",
       "      <td>PMC1345702</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>20181202</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj002</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005</td>\n",
       "      <td>PMC1347365</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>20190104</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005</td>\n",
       "      <td>PMC1347366</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>20181130</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj004</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005</td>\n",
       "      <td>PMC1347367</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>20181123</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj005</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005</td>\n",
       "      <td>PMC1347368</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>20181123</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj006</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005</td>\n",
       "      <td>PMC1347369</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>20181207</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj007</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005</td>\n",
       "      <td>PMC1347370</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20181207</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005</td>\n",
       "      <td>PMC1347371</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20181207</td>\n",
       "      <td>PMC</td>\n",
       "      <td>10.1093/nar/gkj009</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005</td>\n",
       "      <td>PMC1347372</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        date  src                      doi  git_hits  osf_hits  nda_hits  \\\n",
       "0   20181121  PMC   10.1186/1743-8462-2-29         0         0         0   \n",
       "1   20181121  PMC  10.1186/1471-2334-5-116         0         0         0   \n",
       "2   20181121  PMC  10.1186/1471-2105-6-282         0         0         0   \n",
       "3   20181121  PMC       10.1093/nar/gkj431         0         0         0   \n",
       "4   20181121  PMC       10.1093/nar/gkj441         0         0         0   \n",
       "5   20181121  PMC       10.1093/nar/gkj453         0         0         0   \n",
       "6   20181121  PMC       10.1093/nar/gkj461         0         0         0   \n",
       "7   20181124  PMC       10.1093/nar/gkj463         0         0         0   \n",
       "8   20181121  PMC       10.1093/nar/gkj468         0         0         0   \n",
       "9   20181121  PMC       10.1093/nar/gnj008         0         0         0   \n",
       "10  20181112  PMC       10.1093/nar/gnj010         0         0         0   \n",
       "11  20181112  PMC       10.1093/nar/gnj012         0         0         0   \n",
       "12  20181112  PMC       10.1093/nar/gnj015         0         0         0   \n",
       "13  20181202  PMC       10.1093/nar/gkj002         0         0         0   \n",
       "14  20190104  PMC       10.1093/nar/gkj003         0         0         0   \n",
       "15  20181130  PMC       10.1093/nar/gkj004         0         0         0   \n",
       "16  20181123  PMC       10.1093/nar/gkj005         0         0         0   \n",
       "17  20181123  PMC       10.1093/nar/gkj006         0         0         0   \n",
       "18  20181207  PMC       10.1093/nar/gkj007         0         0         0   \n",
       "19  20181207  PMC       10.1093/nar/gkj008         0         0         0   \n",
       "20  20181207  PMC       10.1093/nar/gkj009         0         0         0   \n",
       "\n",
       "    open_neuro  fmri  res    yr       pmcid  idx  \n",
       "0            0     0    0  2005  PMC1345690    0  \n",
       "1            0     0    0  2005  PMC1345691    1  \n",
       "2            0     0    0  2005  PMC1345692    2  \n",
       "3            0     0    0  2006  PMC1345693    3  \n",
       "4            0     0    0  2006  PMC1345694    4  \n",
       "5            0     0    0  2006  PMC1345695    5  \n",
       "6            0     0    0  2006  PMC1345696    6  \n",
       "7            0     0    0  2006  PMC1345697    7  \n",
       "8            0     0    0  2006  PMC1345698    8  \n",
       "9            0     0    0  2006  PMC1345699    9  \n",
       "10           0     0    0  2006  PMC1345700   10  \n",
       "11           0     0    0  2006  PMC1345701   11  \n",
       "12           0     0    0  2006  PMC1345702   12  \n",
       "13           0     0    0  2005  PMC1347365   13  \n",
       "14           0     0    0  2005  PMC1347366   14  \n",
       "15           0     0    0  2005  PMC1347367   15  \n",
       "16           0     0    0  2005  PMC1347368   16  \n",
       "17           0     0    0  2005  PMC1347369   17  \n",
       "18           0     0    0  2005  PMC1347370   18  \n",
       "19           0     0    0  2005  PMC1347371   19  \n",
       "20           0     0    0  2005  PMC1347372   20  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(git_links, columns=['date', 'src', 'doi', 'git_hits', \n",
    "                                      'osf_hits', 'nda_hits', 'open_neuro', \n",
    "                                      'fmri', 'res', 'yr', 'pmcid', 'idx'])\n",
    "df.to_csv('/home/riddleta/ac_knowl/output/bionlp_03.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2975022"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(unique_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2011"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmpyr = '2011'\n",
    "int(tmpyr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "github_ex = re.compile('github')\n",
    "re.search(github_ex, r_out['documents'][0]['passages'][0]['text'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('20181121', 'PMC', 0, 0, 0, 0, 0, 'PMC126257', 350)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "return_useful_data(r) + (j, i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'20181121'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_out['date']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'resolution of airway inflammation and hyperreactivity after in vivo transfer of cd4+cd25+ regulatory t cells is interleukin 10 dependent'"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_out['documents'][0]['passages'][0]['text'].lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_txt = 'resolution of airway inflammation and hyperreactivity after in vivo transfer of cd4+cd25+ regulatory t cells is interleukin 10 dependent. we also posted our data on github.com'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<re.Match object; span=(165, 171), match='github'>"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re.search(re.compile('github'), temp_txt.lower())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = requests.get('https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_json/PMC1350743/unicode')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "r_out = r.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2005'"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_out['documents'][0]['passages'][0]['infons']['year']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'github_ex' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-cfd888f400ac>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mre\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfindall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgithub_ex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpassage\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'text'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'github_ex' is not defined"
     ]
    }
   ],
   "source": [
    "re.findall(github_ex, passage['text'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TITLE\n",
      "ABSTRACT\n",
      "INTRO\n",
      "INTRO\n",
      "INTRO\n",
      "RESULTS\n",
      "RESULTS\n",
      "RESULTS\n",
      "RESULTS\n",
      "RESULTS\n",
      "RESULTS\n",
      "FIG\n",
      "RESULTS\n",
      "RESULTS\n",
      "RESULTS\n",
      "FIG\n",
      "RESULTS\n",
      "RESULTS\n",
      "RESULTS\n",
      "RESULTS\n",
      "RESULTS\n",
      "RESULTS\n",
      "FIG\n",
      "RESULTS\n",
      "RESULTS\n",
      "FIG\n",
      "RESULTS\n",
      "RESULTS\n",
      "FIG\n",
      "DISCUSS\n",
      "DISCUSS\n",
      "DISCUSS\n",
      "DISCUSS\n",
      "DISCUSS\n",
      "DISCUSS\n",
      "DISCUSS\n",
      "DISCUSS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "METHODS\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n",
      "REF\n"
     ]
    }
   ],
   "source": [
    "for passage in r_out['documents'][0]['passages']:\n",
    "    print(passage['infons']['section_type'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'text': 'Resolution of airway inflammation and hyperreactivity after in vivo transfer of CD4+CD25+ regulatory T cells is interleukin 10 dependent',\n",
       " 'offset': 0,\n",
       " 'relations': [],\n",
       " 'infons': {'name_2': 'surname:Robinson;given-names:Douglas S.',\n",
       "  'name_3': 'surname:Lloyd;given-names:Clare M.',\n",
       "  'name_0': 'surname:Kearley;given-names:Jennifer',\n",
       "  'name_1': 'surname:Barker;given-names:Jane E.',\n",
       "  'license': 'This article is distributed under the terms of an Attribution–Noncommercial–Share Alike–No Mirror Sites license for the first six months after the publication date (see http://www.rupress.org/terms). After six months it is available under a Creative Commons License (Attribution–Noncommercial–Share Alike 4.0 Unported license, as described at http://creativecommons.org/licenses/by-nc-sa/4.0/).',\n",
       "  'article-id_publisher-id': '20051166',\n",
       "  'issue': '11',\n",
       "  'article-id_doi': '10.1084/jem.20051166',\n",
       "  'section_type': 'TITLE',\n",
       "  'article-id_pmid': '16314435',\n",
       "  'lpage': '1547',\n",
       "  'article-id_pmc': '1350743',\n",
       "  'fpage': '1539',\n",
       "  'year': '2005',\n",
       "  'type': 'front',\n",
       "  'fn': 'Abbreviations used: AHR, airway hyperreactivity; BAL, bronchoalveolar lavage; Cdyn, dynamic compliance; RL, lung resistance.',\n",
       "  'volume': '202'},\n",
       " 'sentences': [],\n",
       " 'annotations': []}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_out['documents'][0]['passages'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ac_knowl",
   "language": "python",
   "name": "ac_knowl"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
